### PR TITLE
Fix Lua crash if opening door with slab or stair

### PR DESCRIPTION
**Pull request by MarkuBu**

### CODE REVIEW

**Comment by MarkuBu** [comment]
Related https://github.com/minetest/minetest_game/issues/1700

**Comment by MarkuBu** [comment]
Related https://github.com/minetest/minetest/issues/5580

**Comment by paramat** [comment]
Please could you remove `#1700` from the commit message? It links to engine issue 1700 mot MTG.

Although this works, do we understand yet why y is nil? It's better to understand before fixing. This fix may cause strange behaviour elsewhere if we don't understand it.

**Comment by MarkuBu** [comment]
I tried to change the commit message, but it failed. Don't know why. Will try to fix this

I have understand why `y` is nil. This PR fixes the bug with the doors and shouldn't have side effects because it only changes something if `y` is not in the oc table.

Probably there are other nodes in external with a similar effect. Then we have to check again. 

This fix should only patch this bug as long as the engine bug is not fixed. I recommend that this is only a temporary solution

**Comment by paramat** [comment]
Ok. I don't understand how fine pos works so i will not ask for an explanation of why y is nil.

**Comment by MarkuBu** [comment]
> I don't understand how fine pos works

This is pos_off if I point to a normal node (Two examples). Only one axis is ~= 0

```
{
	y = 0,
	x = 0,
	z = -0.5
}
{
	y = 0,
	x = 0.5,
	z = 0
}
```
This is pos_off if I point to the upper half of a door. Two values are ~= 0

```
{
	y = 0.5,
	x = 0,
	z = -0.5
}

```


The for loop loops over x, y and z in pos_off. If everything works as expected only one of this values should be ~= 0. The 0 values are written in the oc list, the other axis is written to nc.

But if two axis are ~= 0 than y is written to nc first and will be overwritten by the next axis

My patch just checks if the y axis is missing in oc and adds it if missing

But I found a bug in my patch. It has no effect, but I want to fix it first

File: builtin/common/misc_helpers.lua [hunk 1]
```diff
@@ -713,6 +713,11 @@ function core.pointed_thing_to_face_pos(placer, pointed_thing)
 			nc = c
 		end
 	end
+
+	if not oc["y"] then
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua]: `not oc["y"]` will always be true

Comment (MarkuBu) [inline builtin/common/misc_helpers.lua]: I know, that's the bug I mentioned before

Comment (MarkuBu) [inline builtin/common/misc_helpers.lua]: And there is another problem I want to fix

**Comment by paramat** [comment]
:+1: (if tested).

**Comment by HybridDog** [comment]
lt looks like pointed_thing.above is not next to pointed_thing.under.
l assume someone could modify the client to add -4 to pointed_thing.above's axes and then join and crash lots of servers.

**Comment by MarkuBu** [comment]
hacked clients are the pest. Let me try to make this bullet proof, if that is a problem. But hacked clients are always a problem

**Comment by MarkuBu** [comment]
I think this will do it. I tested it with manipulated x, y and z coordinates in under and above. The only thing that happens is that nodes can be placed somewhere else. Don't know if that could be a problem (place nodes in protected areas...)

**Comment by HybridDog** [comment]
Just change the for loop (line 709):
```
 	for c, v in pairs(pos_off) do
 		if nc or v == 0 then
 			oc[#oc + 1] = c
 		else
 			offset = v
 			nc = c
 		end
 	end
```

**Comment by MarkuBu** [comment]
I'll test it

**Comment by MarkuBu** [comment]
Simple is sometimes better.

**Comment by paramat** [comment]
:+1: since tested.

**Comment by SmallJoker** [comment]
Tested it too; works. :+1:

**Comment by paramat** [comment]
efd509f796a4870243ce97012aa27940993f4a79

**Pull request closed without merging**